twilight-gateway/src/shard.rs

```diff
@@ -1219,70 +717,245 @@ impl<Q: Queue> Shard<Q> {
+            ShardState::FatallyClosed => {
+                _ = ready!(Pin::new(
+                    self.connection
+                        .as_mut()
+                        .expect("poll_next called after Poll::Ready(None)")
+                )
+                .poll_close(cx));
+                self.connection = None;
+                return Poll::Ready(None);
             }
-        );
+            ShardState::Disconnected { reconnect_attempts } if self.connection.is_none() => {
+                if self.connection_future.is_none() {
+                    let base_url = self
+                        .resume_url
+                        .as_deref()
+                        .or_else(|| self.config.proxy_url())
+                        .unwrap_or(GATEWAY_URL);
+                    let uri =
+                        format!("{base_url}/?v={API_VERSION}&encoding=json{COMPRESSION_FEATURES}");
+
+                    tracing::debug!(url = base_url, "connecting to gateway");
+
+                    let tls = self.config.tls.clone();
+                    self.connection_future = Some(ConnectionFuture(Box::pin(async move {
+                        let secs = 2u8.saturating_pow(reconnect_attempts.into());
+                        time::sleep(Duration::from_secs(secs.into())).await;
+                        Ok(ClientBuilder::new()
+                            .uri(&uri)
+                            .expect("URL should be valid")
+                            .limits(Limits::unlimited())
+                            .connector(&tls)
+                            .connect()
+                            .await?
+                            .0)
+                    })));
+                }
 
-        let non_fatal_code = CloseCode::SessionTimedOut as u16;
-        let non_fatal_status = ConnectionStatus::from_close_code(Some(non_fatal_code));
+                match ready!(Pin::new(&mut self.connection_future.as_mut().unwrap().0).poll(cx)) {
+                    Ok(connection) => {
+                        self.connection = Some(connection);
+                        self.connection_future = None;
+                        self.state = ShardState::Identifying;
+                        #[cfg(any(feature = "zlib-stock", feature = "zlib-simd"))]
+                        self.inflater.reset();
+                    }
+                    Err(source) => {
+                        self.resume_url = None;
+                        self.state = ShardState::Disconnected {
+                            reconnect_attempts: reconnect_attempts + 1,
+                        };
+                        return Poll::Ready(Some(Err(ReceiveMessageError {
+                            kind: ReceiveMessageErrorType::Reconnect,
+                            source: Some(Box::new(source)),
+                        })));
+                    }
+                }
+            }
+            _ => {}
+        }
+
+        let message = loop {
+            ready!(self.poll_flush_pending(cx))?;
```

Comment:
I haven't thought too much about this, but it might be possible that the stream is ready for reading, but not writing. In that case I think we should *still* yield the readable item instead of waiting for it to be writable. AKA don't call `ready!` here, it's good enough to try it as often as possible. For that to work, we'd need a queue of pending items since we can have more than one waiting (this is probably also better long-term).

We however need to use `ready!` if the state is disconnected to ensure the close frame gets flushed.

Reply:
I opted against this due to `poll_flush` not indicating when the gateway event is sent and thus not allowing for proper rate limit registration.

Reply:
I'm not sure if I understand you correctly, but `poll_flush` returning `Poll::Ready` should be indicator enough that the event was sent.

Reply:
Sorry, I meant heartbeat registration. When multiple events are queued and flushed together `poll_flush` can return `Poll::Pending` although *some* events have been sent. Additionally maintaining a queue here would add code complexity that doesn't seem worthwhile since you seldomly send events.

Reply:
I still wonder how that'll impact things like chunking all guilds at startup, can imagine that being somewhat slower like this. But yeah we can't track what exactly was flushed and what wasn't.

Reply:
The implementation right now is exactly the same as the previously used `StreamExt::send`.